src/Features/ExternalAccess/HotReload/Api/HotReloadMSBuildWorkspace.cs

```diff
@@ -0,0 +1,237 @@
+
+        // Map new project id to the corresponding old one based on file path and project name (includes TFM), if it exists, and null for added projects.
+        // Deleted projects won't be included in this map.
+        var projectIdMap = projectInfos.ToDictionary(
+            keySelector: static info => info.Id,
+            elementSelector: info => oldProjectIdsByPath.TryGetValue((info.FilePath!, info.Name), out var oldProjectId) ? oldProjectId : null);
+
+        var newSolution = oldSolution;
+
+        foreach (var newProjectInfo in projectInfos)
+        {
+            Contract.ThrowIfNull(newProjectInfo.FilePath);
+
+            var oldProjectId = projectIdMap[newProjectInfo.Id];
+            if (oldProjectId == null)
+            {
+                newSolution = newSolution.AddProject(newProjectInfo);
+                continue;
+            }
+
+            newSolution = newSolution.WithProjectInfo(ProjectInfo.Create(
+                oldProjectId,
+                newProjectInfo.Version,
+                newProjectInfo.Name,
+                newProjectInfo.AssemblyName,
+                newProjectInfo.Language,
+                newProjectInfo.FilePath,
+                newProjectInfo.OutputFilePath,
+                newProjectInfo.CompilationOptions,
+                newProjectInfo.ParseOptions,
+                MapDocuments(oldProjectId, newProjectInfo.Documents),
+                newProjectInfo.ProjectReferences.Select(MapProjectReference),
+                newProjectInfo.MetadataReferences,
+                newProjectInfo.AnalyzerReferences,
+                MapDocuments(oldProjectId, newProjectInfo.AdditionalDocuments),
+                isSubmission: false,
+                hostObjectType: null,
+                outputRefFilePath: newProjectInfo.OutputRefFilePath)
+                .WithAnalyzerConfigDocuments(MapDocuments(oldProjectId, newProjectInfo.AnalyzerConfigDocuments))
+                .WithCompilationOutputInfo(newProjectInfo.CompilationOutputInfo));
+        }
+
+        var result = SetCurrentSolution(newSolution);
+        UpdateReferencesAfterAdd();
+
+        return result;
+
+        ProjectReference MapProjectReference(ProjectReference pr)
+        {
+            // Only C# and VB projects are loaded by the MSBuildProjectLoader, so some references might be missing.
+            // When a new project is added along with a new project reference the old project id is also null.
+            return new(
+                projectId: projectIdMap.TryGetValue(pr.ProjectId, out var oldProjectId) && oldProjectId != null ? oldProjectId : pr.ProjectId,
+                aliases: pr.Aliases,
+                embedInteropTypes: pr.EmbedInteropTypes);
+        }
+
+        ImmutableArray<DocumentInfo> MapDocuments(ProjectId mappedProjectId, IReadOnlyList<DocumentInfo> documents)
+            => documents.Select(docInfo =>
+            {
+                // TODO: can there be multiple documents of the same path in the project?
```

Comment:
Just to document it somewhere, yes that can happen if a file is added twice. Or is added as both a source file and an additional file. None of these are "real" scenarios, but can happen in broken project scenarios. The only expectation is we don't crash. There might be a bug here since maybe if a source file is being added as an additional file we might try to reuse the same ID which would be bad in this case. Maybe file a bug for tracking, if we refactor this later this code might go away.